ICSharpCode.Decompiler/TypeSystem/MetadataModule.cs

```diff
@@ -477,77 +477,91 @@ IMethod ResolveMethodReference(MemberReferenceHandle memberRefHandle, GenericCon
 				signature = memberRef.DecodeMethodSignature(TypeProvider, context);
 			}
 			else
 			{
-				var declaringType = ResolveDeclaringType(memberRef.Parent, context);
-				var declaringTypeDefinition = declaringType.GetDefinition();
-				if (declaringType.TypeArguments.Count > 0)
-				{
-					classTypeArguments = declaringType.TypeArguments;
-				}
-				// Note: declaringType might be parameterized, but the signature is for the original method definition.
-				// We'll have to search the member directly on declaringTypeDefinition.
 				string name = metadata.GetString(memberRef.Name);
-				signature = memberRef.DecodeMethodSignature(TypeProvider,
-					new GenericContext(declaringTypeDefinition?.TypeParameters));
-				if (declaringTypeDefinition != null)
+				var statedDeclaringType = ResolveDeclaringType(memberRef.Parent, context);
+				var statedDeclaringTypeDefinition = statedDeclaringType.GetDefinition();
+
+				// GetAllBaseTypes() returns types ordered from base -> specific. Traverse the list in reverse
+				// to check in the opposite direction (most specific to least specific).
+				var candidateDeclaringTypes = statedDeclaringType.GetAllBaseTypes().ToList();
+				for (int i = candidateDeclaringTypes.Count - 1; i >= 0; i--)
 				{
-					// Find the set of overloads to search:
-					IEnumerable<IMethod> methods;
-					if (name == ".ctor")
-					{
-						methods = declaringTypeDefinition.GetConstructors();
-					}
-					else if (name == ".cctor")
-					{
-						methods = declaringTypeDefinition.Methods.Where(m => m.IsConstructor && m.IsStatic);
-					}
-					else
-					{
-						methods = declaringTypeDefinition.GetMethods(m => m.Name == name, GetMemberOptions.IgnoreInheritedMembers)
-							.Concat(declaringTypeDefinition.GetAccessors(m => m.Name == name, GetMemberOptions.IgnoreInheritedMembers));
-					}
-					// Determine the expected parameters from the signature:
-					ImmutableArray<IType> parameterTypes;
-					if (signature.Header.CallingConvention == SignatureCallingConvention.VarArgs)
-					{
-						parameterTypes = signature.ParameterTypes
-							.Take(signature.RequiredParameterCount)
-							.Concat(new[] { SpecialType.ArgList })
-							.ToImmutableArray();
-					}
-					else
+					var declaringType = candidateDeclaringTypes[i];
+					var declaringTypeDefinition = declaringType.GetDefinition();
+
+					if (declaringType.TypeArguments.Count > 0)
 					{
-						parameterTypes = signature.ParameterTypes;
+						classTypeArguments = declaringType.TypeArguments;
 					}
-					// Search for the matching method:
-					method = null;
-					foreach (var m in methods)
+
+					// Note: declaringType might be parameterized, but the signature is for the original method definition.
+					// We'll have to search the member directly on declaringTypeDefinition.
+					signature = memberRef.DecodeMethodSignature(TypeProvider,
+						new GenericContext(declaringTypeDefinition?.TypeParameters));
```

Comment:
I don't think this deal with `GenericContext` is correct once inheritance comes into play -- the base class may have different type parameters than the derived class.
Consider a class hierarchy like:
```
class B<T> { void Add(T); }
class C<U> : B<List<U>> { }
```
Now I believe an Add() call on `C<int>` can be encoded as a call to `C<U=int>.Add(List<U>)`. This overload won't be found in the declarationTypeDefinition of the base class. In edge cases you might end up returning the wrong overload.

Reply:
Really specialization needs to happen in two steps here:
We need to use the partial specialization `B<List<U>>` for comparing the signature of `Add`, and only then substitute `U=int`.
So you need to go from declaringType to declaringTypeDefinition just once for the declaringType mentioned in the IL, not for each base class separately.